app/src/main/java/crux/bphc/cms/activities/TokenViewModel.kt

```diff
@@ -0,0 +1,223 @@
+package crux.bphc.cms.activities
+
+import android.app.Application
+import android.content.Context
+import android.content.Intent
+import android.net.Uri
+import android.util.Base64
+import android.util.Log
+import androidx.lifecycle.AndroidViewModel
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.viewModelScope
+import crux.bphc.cms.app.MyApplication
+import crux.bphc.cms.app.Urls
+import crux.bphc.cms.app.appendOrSetQueryParameter
+import crux.bphc.cms.helper.CourseDataHandler
+import crux.bphc.cms.helper.CourseRequestHandler
+import crux.bphc.cms.models.SingleLiveEvent
+import crux.bphc.cms.models.UserAccount
+import crux.bphc.cms.models.core.UserDetail
+import crux.bphc.cms.models.course.Module
+import crux.bphc.cms.network.APIClient
+import crux.bphc.cms.network.MoodleServices
+import crux.bphc.cms.utils.UserUtils
+import crux.bphc.cms.utils.Utils
+import io.realm.Realm
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+import java.security.NoSuchAlgorithmException
+import java.util.*
+
+class TokenViewModel(application: Application) : AndroidViewModel(application) {
+
+    private val TAG = javaClass.simpleName
+
+    private val moodleServices: MoodleServices = APIClient.getRetrofitInstance().create(MoodleServices::class.java)
+
+    private val _onErrorMessage: MutableLiveData<SingleLiveEvent<String>> = MutableLiveData()
```

Comment:
Do we really need to prepend private variables with an `_`?

Reply:
I am a little confused myself about prepending private variables with an `_` too, but again, from the convention I've seen, the LiveData is always the MutableLiveData's variable name prepended with an `_`. Since only the ViewModel changes the value of the MutableLiveData and the Fragment should only be able to read LiveData, MutableLiveData is private and has the `_` before its variable name.

Reply:
I guess it is fine in this case since the `MutableLiveData` and `LiveData` have the same name and both of them are required, one for pushing data and for observing.